design-docs/configuration-dsl.md

```diff
@@ -0,0 +1,376 @@
+are debug optimized and final. The software model, as well as Android, have this dimension as a first class citizen.
+Android typed the items inside their buildTypes container instead of relying on `ext` properties. With the new plugins, the build types will
+be typed as well.
+
+## Open Issues
+
+## DSL
+```
+apply plugin: 'build-types'
+
+buildTypes {
+    debug(MyDebugType)
+    release(BuildType)
+}
+```
+
+## Implementations
+```
+// Common build type
+public interface BuildType extends Named {}
+
+// Container type to hold build types
+public interface BuildTypeContainer extends ExtensiblePolymorphicDomainObjectContainer<BuildType> {}
+```
+
+# Linkage
+
+The linkage container specifies how you want to link your application/library/module together. The static and shared library are simple variants of
+the binary.
+
+## Open Issues
+
+* Do we want to specialize the linkage type? For example, `static(MyCustomLinkage)`?
+
+## DSL
+```
+apply plugin: 'linkage'
+
+linkages {
+    // Static library
+    static
+    // Shared library binary
+    shared
+    // Compile sources together
+    source
+    // Executable binary
+    executable
+}
+```
+
+## Implementations
+```
+public interface Linkage extends Named {}
+public interface LinkageContainer extends NamedDomainObjectContainer<Linkage> {}
+```
+
+# Prebuilt Libraries
+All native dependencies will be using Gradle's dependency management. Due to the complexity of native prebuilt library support, Gradle
+will provide a public API for declaring them.
+
+## Open Issues
+* Should Gradle allow prebuilt libraries as custom DSL or virtual repository (see example)?
+
+## DSL
+Virtual Repository
+```
+apply plugin: 'boost-cpp'
+
+repositories {
+    boost {
+        // Configuration allowed by the Boost plugin
+    }
+}
+
+dependencies {
+    // This dependency will ask the `boost` virtual repository where to find Boost:ASIO version 1.63. The repository can look on disk,
+    // at a remote location or event checkout the code and build from source.
+    implementation group: 'org.boost', name: 'asio', version: '1.63'
+}
+```
+
+Custom DSL
+```
+apply plugin: 'boost-cpp'
+
+dependencies {
+    // This dependency will provide the right dependency to find Boost:ASIO version 1.63. The dependency element can look on disk,
+    // at a remote location or event checkout the code and build from source.
+    implementation boost('1.63').asio
+}
+```
+
+
+# Instruction Set Architecture (ISA)
+The architecture is the interface between the software and hardware. Software written for an ISA can run on different implementations of the same ISA.
+
+## Open Issues
+* Do we want to specialize the architecture type? For example, `snesArch(Snes)`.
+
+## DSL
+```
+architectures {
+    x64 {
+        pointerSize = 8
+        endianness = 'little'
+    }
+    x86
+}
+```
+
+# Operating System
+```
+operatingSystems {
```

Comment:
There are 2 separate aspects to OS:

1. The aspects that need to be baked into the binaries in order for them to run on the target OS (the file type, entry points, dynamic linker instructions, ABI used by the system libraries, etc).
2. The system APIs that the project's code uses, which need to be available at compile/link/debug/run time.

We shouldn't bundle these things together, as a binary may be built so that it runs on a particular OS version but does not use (directly) anything other than the standard language library. The reverse is probably true as well, eg I can use the linux API and be built to run on windows (via cygwin for example).